rust/lance/src/index/prefilter.rs

```diff
@@ -109,17 +111,72 @@ impl DatasetPreFilter {
         for frag_id in missing_frags.into_iter() {
             deleted_ids.insert_fragment(frag_id);
         }
-        Ok(Arc::new(deleted_ids))
+        Ok(Arc::new(RowIdMask::from_block(deleted_ids)))
     }
 
-    /// Creates a task to load deleted row ids in `fragments`
+    #[instrument(level = "debug", skip_all)]
+    async fn do_create_deletion_mask_row_id(dataset: Arc<Dataset>) -> Result<Arc<RowIdMask>> {
+        // This can only be computed as an allow list, since we have no idea
+        // what the row ids were in the missing fragments.
+
+        // For each fragment, compute which row ids are still in use.
+        let dataset_ref = dataset.as_ref();
+
+        let row_ids_and_deletions = stream::iter(dataset.get_fragments())
+            .map(|frag| async move {
+                let row_ids = load_row_id_sequence(dataset_ref, frag.metadata());
+                let deletion_vector = frag.get_deletion_vector();
+                let (row_ids, deletion_vector) = join!(row_ids, deletion_vector);
+                Ok::<_, crate::Error>((row_ids?, deletion_vector?))
+            })
+            .buffer_unordered(10)
+            .try_collect::<Vec<_>>()
+            .await?;
+
+        // The process of computing the final mask is CPU-bound, so we spawn it
+        // on a blocking thread.
+        let allow_list = tokio::task::spawn_blocking(move || {
+            row_ids_and_deletions.into_iter().fold(
+                RowIdTreeMap::new(),
+                |mut allow_list, (row_ids, deletion_vector)| {
+                    let row_ids = if let Some(deletion_vector) = deletion_vector {
+                        // We have to mask the row ids
+                        row_ids.as_ref().iter().enumerate().fold(
+                            RowIdTreeMap::new(),
+                            |mut allow_list, (idx, row_id)| {
+                                if !deletion_vector.contains(idx as u32) {
+                                    allow_list.insert(row_id);
+                                }
+                                allow_list
+                            },
+                        )
+                    } else {
+                        // Can do a direct translation
+                        RowIdTreeMap::from(row_ids.as_ref())
+                    };
+                    allow_list |= row_ids;
+                    allow_list
+                },
+            )
+        })
+        .await?;
+
+        Ok(Arc::new(RowIdMask::from_allowed(allow_list)))
+    }
+
+    /// Creates a task to load mask to filter out deleted rows.
+    ///
+    /// Sometimes this will be a block list of row ids that are deleted, based
+    /// on the deletion files in the fragments. If stable row ids are used and
+    /// there are missing fragments, this may instead be an allow list, since
+    /// we can't easily compute the block list.
```

Comment:
Hmm...if it is an allow list I think this may have some performance impact on prefiltering?

Reply:
This is possible. I will be doing further benchmarking to understand this later.